numhask/src/NumHask/Analysis/Metric.hs

```diff
@@ -0,0 +1,366 @@
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- instance Metric Int8 Int8 where
+--   distanceL1 a b = normL1 (a - b)
+--   distanceL2 a b = normL2 (a - b)
+--   distanceLp p a b = normLp p (a - b)
+
+-- instance Metric Int16 Int16 where
+--   distanceL1 a b = normL1 (a - b)
+--   distanceL2 a b = normL2 (a - b)
+--   distanceLp p a b = normLp p (a - b)
+
+-- instance Metric Int32 Int32 where
+--   distanceL1 a b = normL1 (a - b)
+--   distanceL2 a b = normL2 (a - b)
+--   distanceLp p a b = normLp p (a - b)
+
+-- instance Metric Int64 Int64 where
+--   distanceL1 a b = normL1 (a - b)
+--   distanceL2 a b = normL2 (a - b)
+--   distanceLp p a b = normLp p (a - b)
+
+-- -- fixme: circular distance may be more appropriate
+-- instance Metric Word Word where
+--   distanceL1 a b = fromInteger $ normL1 (toInteger a - toInteger b)
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- instance Metric Word8 Word8 where
+--   distanceL1 a b = fromInteger $ normL1 (toInteger a - toInteger b)
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- instance Metric Word16 Word16 where
+--   distanceL1 a b = fromInteger $ normL1 (toInteger a - toInteger b)
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- instance Metric Word32 Word32 where
+--   distanceL1 a b = fromInteger $ normL1 (toInteger a - toInteger b)
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- instance Metric Word64 Word64 where
+--   distanceL1 a b = fromInteger $ normL1 (toInteger a - toInteger b)
+--   distanceL2 a b = fromInteger $ normL2 (toInteger a - toInteger b)
+--   distanceLp p a b = fromInteger (normLp (toInteger p) (toInteger a - toInteger b))
+
+-- | todo: This should probably be split off into some sort of alternative Equality logic, but to what end?
+class (Eq a, Unital (Add a)) =>
+      Epsilon a where
+  nearZero :: a -> Bool
+  nearZero a = a == zero
+
+  aboutEqual :: a -> a -> Bool
+  default aboutEqual :: (Group (Add a), Commutative (Add a)) => a -> a -> Bool
+  aboutEqual a b = nearZero $ a - b
+
+  positive :: (Signed a) => a -> Bool
+  positive a = a == abs a
+  veryPositive :: (Signed a) => a -> Bool
+  veryPositive a = P.not (nearZero a) && positive a
+  veryNegative :: (Signed a) => a -> Bool
+  veryNegative a = P.not (nearZero a P.|| positive a)
+
+infixl 4 ≈
+
+-- | todo: is utf perfectly acceptable these days?
+(≈) :: (Epsilon a) => a -> a -> Bool
```

Comment:
I prefer plain ascii operators such as `(~=)`

Reply:
I also avoid Unicode in my code. I do use [Hasklig](https://github.com/i-tu/Hasklig/) to make things prettier but doesn't affect how anyone else sees the code. Sadly, it doesn't not currently contain a ligature for ~=.
Since there is a lot going on in this PR already, I suggest sticking to ascii to remove this as a topic of debate.